### PR TITLE
Support batched training

### PR DESCRIPTION
**Pull request by ylxdzsw**
This PR add batched training by:
- adding a `Batched` type, which wraps an Iterator and output in batch
- add keyword argument `all` for `update!` of MXNet backend, which copy the final params to all execs
- make `mse!` supports batched arguments
- alter the training process of TensorFlow backend
- fix a little bug of `CatMat`

### CODE REVIEW

File: src/backend/tensorflow/model.jl [hunk 1]
```diff
@@ -57,11 +57,11 @@ function Flux.train!(m::Model, train, test=[]; epoch = 1, η = 0.1,
...
   for e in 1:epoch
     info("Epoch $e\n")
-    @progress for (x, y) in train
+    @progress for (x, y) in Batched(train, batch)
       y, cur_loss, _ = run(m.session, vcat(m.output, Loss, minimize_op),
-                           Dict(m.inputs[1] => batchone(convertel(Float32, x)),
-                                Y => batchone(convertel(Float32, y))))
-      if i % 5000 == 0
+                           Dict(m.inputs[1] => rebatch(convertel(Float32, rawbatch(x))),
+                                Y => rebatch(convertel(Float32, rawbatch(y)))))
```
Comment (MikeInnes) [inline src/backend/tensorflow/model.jl]: `convertel` should already work on batches without needing to convert to an array and back

File: src/cost.jl [hunk 2]
```diff
@@ -1,7 +1,7 @@
 export mse, mse!
 
 function mse!(Δ, pred, target)
-  map!(-, Δ, pred, target)
+  map!(-, Δ, rawbatch(pred), rawbatch(target))
```
Comment (MikeInnes) [inline src/cost.jl]: It seems wrong that `mse!` needs to know about batching; in general, we should be able to write functions over generic arrays and have batching work for free. In this case, perhaps we can implement `broadcast!` for `Batch` so that `mse!` works as is? If it's complicated to do this it doesn't have to block the PR.

File: src/dims/batching.jl [hunk 3]
```diff
@@ -59,3 +62,71 @@ runrawbatched(f, xs...) =
...
+mutable struct Batched{T,S}
+  batch::Int
+  iter::T
+  "`Batched` always read a batch in advance, and store it in `buf`"
+  buf::S
+  i
+end
+
+function Batched(iter::T, batch::Integer) where T
+  batch >= 1 || throw(ArgumentError("batch size must >= 1"))
```
Comment (MikeInnes) [inline src/dims/batching.jl]: Is there a particular reason to disable batch size 1?

File: src/dims/batching.jl [hunk 4]
```diff
@@ -59,3 +62,71 @@ runrawbatched(f, xs...) =
...
+
+next(x::Batched, ::Bool) = x.buf, false
+
+# will be less hacky if https://github.com/JuliaLang/julia/issues/18823
+function done(x::Batched, fresh)
+  fresh && return false
+
+  for ibatch in 1:x.batch
+    if done(x.iter, x.i)
+      warn("cannot perfectly divide data by batch size, remainder will be discarded")
```
Comment (MikeInnes) [inline src/dims/batching.jl]: In principle, we could just return a smaller final batch. I'm not sure if we should or not, but it might be a nicer default.

File: src/utils.jl [hunk 5]
```diff
@@ -6,16 +6,20 @@ initn(dims...) = randn(dims...)/100
 
 function train!(m, train, test = []; epoch = 1, batch = 10, η = 0.1)
     i = 0
-    Δ = zeros(length(train[1][2]))
+    Δ = zeros(batch, length(train[1][2]))
     for _ in 1:epoch
-      @progress for (x, y) in train
+      @progress for (x, y) in Batched(train, batch)
```
Comment (MikeInnes) [inline src/utils.jl]: I think perhaps `train!` should stay generic and continue to accept an iterator of either raw arrays or batches, rather than trying to handle batching entirely itself. That gives you the control to batch things ahead of time, for example.

File: src/utils.jl [hunk 6]
```diff
@@ -6,16 +6,20 @@ initn(dims...) = randn(dims...)/100
...
         pred = m(x)
-        any(isnan, pred) && error("NaN")
+        any(any(isnan, p) for p in pred) && error("NaN")
         err = mse!(Δ, pred, y)
         back!(m, Δ, x)
-        i % batch == 0 && update!(m, η)
-        i % 1000 == 0 && @show accuracy(m, test)
+        if i % 100 == 0
+          update!(m, η, all=true)
+          @show accuracy(m, test)
```
Comment (MikeInnes) [inline src/utils.jl]: When I last tried this I had problems with `accuracy` not taking batches. Are the `mse!` changes enough to fix that?

File: src/backend/mxnet/model.jl [hunk 7]
```diff
@@ -127,7 +127,17 @@ function Flux.back!(m::Model, Δ, xs...)
   end
 end
 
-Flux.update!(m::Model, η) = (update!(m.last, η); m)
+Flux.update!(m::Model, η; all=false) = begin
```
Comment (ylxdzsw) [inline src/backend/mxnet/model.jl]: Yes, I saw your note in the file. I will try to implement that way. Just another idea: is it possible to use exactly the same `NDArray` among `Exec`s? I think move the `params` local varaiable in `mxparams` function to the `Model` or `Graph` object and do not reinitialize it every time should do the trick. This should completely get rid of this problem and in addition, save precious GPU memory.

File: src/backend/tensorflow/model.jl (hunk 1 above)
Comment (ylxdzsw) [inline src/backend/tensorflow/model.jl]: Yes, it does. I didn't notice that.

File: src/cost.jl (hunk 2 above)
Comment (ylxdzsw) [inline src/cost.jl]: I'm afraid that overloading `broadcast!` is not enough, as this is the `map!` in julia 0.6:

```
function map!{F}(f::F, dest::AbstractArray, A::AbstractArray, B::AbstractArray)
    for (i, j, k) in zip(eachindex(dest), eachindex(A), eachindex(B))
        dest[i] = f(A[j], B[k])
    end
    return dest
end
```

We treat `Batch` as an `Vector` of `Array`s, so the dimensions of those `eachindex`s are not match. What about just specialize `mse!` for `Batch`s?

File: src/dims/batching.jl (hunk 3 above)
Comment (ylxdzsw) [inline src/dims/batching.jl]: No, it doesn't. batch == 1 is allowed and I tested.

File: src/dims/batching.jl (hunk 4 above)
Comment (ylxdzsw) [inline src/dims/batching.jl]: AFAIK most frameworks simply discard the remainder, but yes, it worth being implemented as at least an option for user to choose.

File: src/utils.jl (hunk 5 above)
Comment (ylxdzsw) [inline src/utils.jl]: How do we know the elements of an iterator is a `Batch` or raw `Array`? For example, if the iterator yields `Array`s of size (5, 28, 28, 3), is it a `Batch` of (28, 28, 3) `Array`s or just an element of 4-dimension array? Or should we export `rebatch` function and tell users to pack `Batch`es in the first case?

Comment (ylxdzsw) [inline src/utils.jl]: Here's the Keras way: Keras has two methods, `fit` and `fit_generator`, the `fit` accept a big array (can be treat as an entire batch of all data) and split it by `batch_size`; while `fit_generator` accepts iterator that generates batchs and do not have `batch_size` argument. I think we can do the same thing: `train!` still accept `batch` and pack the elements of iterator accroding to it, and add a new `train_batched!` allow user to yield the batch and we just map `rebatch` on the iterator.

File: src/utils.jl (hunk 6 above)
Comment (ylxdzsw) [inline src/utils.jl]: umm, In fact I didn't touch it and still evaluate the model one sample at a time. It should not be hard to support batch though.

File: src/dims/batching.jl (hunk 4 above)
Comment (MikeInnes) [inline src/dims/batching.jl]: Ok, I'm happy to follow precedent here.

File: src/dims/batching.jl (hunk 3 above)
Comment (MikeInnes) [inline src/dims/batching.jl]: Obviously – reviewing PRs too late :)

File: src/utils.jl (hunk 6 above)
Comment (MikeInnes) [inline src/utils.jl]: Whoops, my bad, I was thinking of when I tried `train!` with an iterator of batches; which almost worked aside from this line. Of course, if you expect `test` to be samples then this will work. We can discuss that assumption elsewhere.

File: src/backend/mxnet/model.jl (hunk 7 above)
Comment (MikeInnes) [inline src/backend/mxnet/model.jl]: I don't know that it *won't* work, so I guess it's worth a shot :)

File: src/cost.jl (hunk 2 above)
Comment (MikeInnes) [inline src/cost.jl]: We can just use `broadcast!` instead of `map!` here.

File: src/utils.jl (hunk 5 above)
Comment (MikeInnes) [inline src/utils.jl]: Yes, `Batch` is supposed to be user facing. The idea is that we then don't ever have to worry about the question of "is this a batch?", making things less error prone. We should also be able to write functions that work in a generic way for batches and samples, avoiding duplication.

The model interface will accept either a sample or a batch, so in theory `train!` need not know about batching at all. We can then call either `train!(m, xs)` or `train!(m, batched(xs, 10))`. (This doesn't prevent `train!` from having a `batch_size` argument for convenience / familiarity of course.)

Comment (MikeInnes) [inline src/utils.jl]: #21 will invalidate some of what I said above. So let's just get this working in a simple way, as you've done it, and worry about making things more general later.

**Comment by ylxdzsw** [comment]
Rebased, plus all execs in the same mx model now share the same set of args and grads - we don't need to worry about `last` anymore.

**Comment by ylxdzsw** [comment]
Rebased with latest master again. Will this get merged in the near future? Keeping syncing with master is somewhat error prone.

**Comment by MikeInnes** [comment]
Err, whoops. I tried to push a change to your branch but it ended up deleting the commits.

Rest assured I have these locally and I'm rebasing/merging them right now.

**Comment by ylxdzsw** [comment]
Yeah I have a local backup too. I'm a little surprised that GitHub allows pushing to forked repos directly.

**Comment by MikeInnes** [comment]
Maintainers can push to branches used for PRs, which is pretty handy for patching things up before a merge without going back and forth. Of course, as soon as the PR closed itself that no longer applied, which meant I couldn't fix my mistake :P
